### PR TITLE
Spring boot security

### PR DESCRIPTION
**Pull request by Yam0r**

### CODE REVIEW

File: src/main/java/mate/academy/springintro/controller/AuthenticationController.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+@RequiredArgsConstructor
+@RestController
+@RequestMapping("/auth")
+public class AuthenticationController {
+    private final UserService userService;
+    private final AuthenticationService authenticationService;
+
+    @PostMapping("/registration")
+    public UserResponseDto registerUser(@RequestBody UserRegistrationRequestDto requestDto)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/controller/AuthenticationController.java]:
```suggestion
    public UserResponseDto registerUser(@RequestBody @Valid UserRegistrationRequestDto requestDto)
```

File: src/main/java/mate/academy/springintro/dto/user/UserRegistrationRequestDto.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+import jakarta.validation.constraints.Email;
+import jakarta.validation.constraints.NotBlank;
+import lombok.Data;
+import org.hibernate.validator.constraints.Length;
+
+@Data
+public class UserRegistrationRequestDto {
+    @NotBlank
+    @Email
+    private String email;
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/dto/user/UserRegistrationRequestDto.java:14]: there are should be more fields 
Let's look at task description

{
  "email": "john.doe@example.com",
  "password": "securePassword123",
  "repeatPassword": "securePassword123",
  "firstName": "John",
  "lastName": "Doe",
  "shippingAddress": "123 Main St, City, Country"
}

File: src/main/java/mate/academy/springintro/controller/AuthenticationController.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+            throws RegistrationException {
+        return userService.register(requestDto);
+    }
+
+    @PostMapping("/login")
+    public boolean login(@RequestBody UserLoginRequestDto requestDto) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/controller/AuthenticationController.java]: You don't need to implement /login endpoint in this HW! (see the task description)

File: src/main/java/mate/academy/springintro/controller/AuthenticationController.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+import mate.academy.springintro.service.UserService;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RequiredArgsConstructor
+@RestController
+@RequestMapping("/auth")
+public class AuthenticationController {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/controller/AuthenticationController.java:20]: add swagger annotations to controller

File: src/main/java/mate/academy/springintro/dto/user/UserResponseDto.java [hunk 5]
```diff
@@ -0,0 +1,9 @@
+package mate.academy.springintro.dto.user;
+
+import lombok.Data;
+
+@Data
+public class UserResponseDto {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/dto/user/UserResponseDto.java:6]: example of response 
{
  "id": 1,
  "email": "john.doe@example.com",
  "firstName": "John",
  "lastName": "Doe",
  "shippingAddress": "123 Main St, City, Country"
}

File: src/main/java/mate/academy/springintro/exception/RegistrationException.java [hunk 6]
```diff
@@ -0,0 +1,8 @@
+package mate.academy.springintro.exception;
+
+public class RegistrationException extends Exception {
+
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/exception/RegistrationException.java:12]: remove empty line

File: src/main/java/mate/academy/springintro/mapper/UserMapper.java [hunk 7]
```diff
@@ -0,0 +1,8 @@
+package mate.academy.springintro.mapper;
+
+import mate.academy.springintro.dto.user.UserResponseDto;
+import mate.academy.springintro.model.User;
+
+public interface UserMapper {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/mapper/UserMapper.java:10]:
```suggestion
@Mapper(config = MapperConfig.class)
public interface UserMapper {
```

File: src/main/java/mate/academy/springintro/model/User.java [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+import jakarta.persistence.Entity;
+import jakarta.persistence.GeneratedValue;
+import jakarta.persistence.GenerationType;
+import jakarta.persistence.Id;
+import jakarta.persistence.Table;
+import lombok.Data;
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.Where;
+
+@Data
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/model/User.java]:
```suggestion
@Getter
@Setter
```

File: src/main/java/mate/academy/springintro/model/User.java [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+@Entity
+@SQLDelete(sql = "UPDATE users SET is_deleted = true WHERE id=?")
+@Where(clause = "is_deleted = false")
+@Table(name = "users")
+public class User {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/model/User.java:20]: Add User entity with the following fields:
id (Long, PK)
email (String, not null, unique)
password (String, not null)
firstName (String, not null)
lastName (String, not null)
shippingAddress (String)

File: src/main/java/mate/academy/springintro/security/AuthenticationService.java [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+import mate.academy.springintro.model.User;
+import mate.academy.springintro.repository.UserRepository;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+public class AuthenticationService {
+    private final UserRepository userRepository;
+
+    public boolean authenticate(UserLoginRequestDto requestDto) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/security/AuthenticationService.java]: remove method

File: src/main/java/mate/academy/springintro/service/impl/UserMapperImpl.java [hunk 11]
```diff
@@ -0,0 +1,17 @@
+package mate.academy.springintro.service.impl;
+
+import mate.academy.springintro.dto.user.UserResponseDto;
+import mate.academy.springintro.mapper.UserMapper;
+import mate.academy.springintro.model.User;
+import org.springframework.stereotype.Component;
+
+@Component
+public class UserMapperImpl implements UserMapper {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/service/impl/UserMapperImpl.java]: don't implement mapper

File: src/main/java/mate/academy/springintro/service/impl/UserServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+@Service
+@RequiredArgsConstructor
+public class UserServiceImpl implements UserService {
+    private final UserRepository userRepository;
+    private final UserMapper userMapper;
+
+    @Override
+    public UserResponseDto register(UserRegistrationRequestDto requestDto)
+            throws RegistrationException {
+        if (userRepository.findByEmail(requestDto.getEmail()).isPresent()) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/service/impl/UserServiceImpl.java]: it's better to use 
```suggestion
        if (userRepository.existsByEmail(requestDto.getEmail())) {
```

File: src/main/java/mate/academy/springintro/service/impl/UserServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+            throw new RegistrationException("Can't register user");
+        }
+        User user = new User();
+        user.setEmail(requestDto.getEmail());
+        user.setPassword(requestDto.getPassword());
+        User savedUser = userRepository.save(user);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/service/impl/UserServiceImpl.java]:
```suggestion
        User user = userMapper.toUser(request);
        userRepository.save(user);
        return userMapper.toUserResponse(user);
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Add FieldMatch annotation for checking that password and repeatPassword field in the UserRegistrationRequestDto are the same
Consider to handle the RegistrationException using the GlobalExceptionHandler

File: pom.xml [hunk 14]
```diff
@@ -114,6 +114,13 @@
 			<groupId>org.hibernate.validator</groupId>
 			<artifactId>hibernate-validator</artifactId>
 		</dependency>
+
+		<dependency>
+			<groupId>commons-beanutils</groupId>
+			<artifactId>commons-beanutils</artifactId>
+			<version>1.9.4</version>
+		</dependency>
```
Comment (liliia-ponomarenko) [inline pom.xml]: Why do you need this dependency?

File: src/main/java/mate/academy/springintro/controller/AuthenticationController.java [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@Tag(name = "Authentication", description = "Endpoints for user authentication and registration")
+@RequiredArgsConstructor
+@RestController
+@RequestMapping("/auth")
+public class AuthenticationController {
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/springintro/controller/AuthenticationController.java:20]: Let's remove the commented code

File: src/main/java/mate/academy/springintro/exception/FieldMatchValidator.java [hunk 16]
```diff
@@ -0,0 +1,28 @@
+package mate.academy.springintro.exception;
+
+import jakarta.validation.ConstraintValidator;
+import jakarta.validation.ConstraintValidatorContext;
+import org.apache.commons.beanutils.BeanUtils;
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/springintro/exception/FieldMatchValidator.java]: okay now I see that you use it/ But I recommend you use org.springframework.beans.BeanWrapperImpl

File: src/main/java/mate/academy/springintro/exception/FieldMatchValidator.java [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+    }
+
+    @Override
+    public boolean isValid(Object value, ConstraintValidatorContext context) {
+        try {
+            String firstValue = BeanUtils.getProperty(value, firstFieldName);
+            String secondValue = BeanUtils.getProperty(value, secondFieldName);
+
+            return firstValue != null && firstValue.equals(secondValue);
+        } catch (Exception e) {
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/springintro/exception/FieldMatchValidator.java]: What exception are you expecting here? Let’s  remove try-catch  block

File: src/main/java/mate/academy/springintro/security/AuthenticationService.java [hunk 18]
```diff
@@ -0,0 +1,11 @@
+package mate.academy.springintro.security;
+
+import lombok.RequiredArgsConstructor;
+import mate.academy.springintro.repository.UserRepository;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+public class AuthenticationService {
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/springintro/security/AuthenticationService.java]: Let’s  remove the service without the realization

File: src/main/java/mate/academy/springintro/service/impl/UserServiceImpl.java [hunk 19]
```diff
@@ -0,0 +1,29 @@
...
+
+@Service
+@RequiredArgsConstructor
+public class UserServiceImpl implements UserService {
+    private final UserRepository userRepository;
+    private final UserMapper userMapper;
+
+    @Override
+    public UserResponseDto register(UserRegistrationRequestDto requestDto)
+            throws RegistrationException {
```
Comment (liliia-ponomarenko) [inline src/main/java/mate/academy/springintro/service/impl/UserServiceImpl.java:21]: I can't find handling RegistrationException. Let’s add this logic)

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let’s fix a few mistakes ;)

File: src/main/java/mate/academy/springintro/model/User.java [hunk 20]
```diff
@@ -0,0 +1,36 @@
...
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @Column(nullable = false, unique = true)
+    private String email;
+    @Column(nullable = false)
+    private String password;
+    @Column(name = "is_deleted", nullable = false)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/model/User.java]:
```suggestion
    @Column(nullable = false)
```

File: src/main/java/mate/academy/springintro/model/User.java [hunk 21]
```diff
@@ -0,0 +1,36 @@
...
+    private boolean isDeleted;
+    @Column(nullable = false)
+    private String firstName;
+    @Column(nullable = false)
+    private String lastName;
+
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/model/User.java]:
```suggestion
```

File: src/main/java/mate/academy/springintro/security/AuthenticationService.java (hunk 18 above)
Comment (Elena-Bruyako) [reply src/main/java/mate/academy/springintro/security/AuthenticationService.java]: fix comment

File: src/main/java/mate/academy/springintro/service/impl/UserServiceImpl.java (hunk 19 above)
Comment (Elena-Bruyako) [reply src/main/java/mate/academy/springintro/service/impl/UserServiceImpl.java:21]: fix comment

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
fix previous comments

File: src/main/java/mate/academy/springintro/model/User.java [hunk 22]
```diff
@@ -0,0 +1,35 @@
...
+import lombok.Getter;
+import lombok.Setter;
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.Where;
+
+@Getter
+@Setter
+@Entity
+@SQLDelete(sql = "UPDATE users SET is_deleted = true WHERE id=?")
+@Where(clause = "is_deleted = false")
```
Comment (okuzan) [inline src/main/java/mate/academy/springintro/model/User.java]: where is outdated, use SQLRestriction

Comment (okuzan) [inline src/main/java/mate/academy/springintro/model/User.java]:
```suggestion
@SQLDelete(sql = "UPDATE users SET is_deleted = TRUE WHERE id = ?")
@Where(clause = "is_deleted = FALSE")
```

**Review by okuzan — CHANGES_REQUESTED** [review]

File: src/main/java/mate/academy/springintro/exception/CustomGlobalExceptionHandler.java [hunk 23]
```diff
@@ -42,4 +43,12 @@ private String getErrorMessage(ObjectError e) {
         }
         return e.getDefaultMessage();
     }
+
+    @ExceptionHandler(RegistrationException.class)
+    public ResponseEntity<Object> handleRegistrationException(RegistrationException ex) {
+        Map<String, Object> body = new LinkedHashMap<>();
+        body.put("timestamp", LocalDateTime.now());
+        body.put("message", ex.getMessage());
+        return new ResponseEntity<>(body, HttpStatus.BAD_REQUEST);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springintro/exception/CustomGlobalExceptionHandler.java]:
```suggestion
        return new ResponseEntity<>(body, HttpStatus.CONFLICT);
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
one comment

**Review by okuzan — APPROVED** [review]

**Pull request merged**